BayeuxClient.agent.lib.nut

```diff
@@ -0,0 +1,961 @@
+            _send([message], callback);
+        } else {
+            _messageQueue.push([message, callback]);
+        }
+    }
+
+    function _processQueue() {
+        // We have a free request
+        if (!_isDisconnecting && _extraHttpRequest == null && _messageQueue.len() > 0) {
+            // msgCb is an array [<message>, <callback>]
+            local msgCb = _messageQueue.remove(0);
+            _send([msgCb[MESSAGE_INDEX]], msgCb[CALLBACK_INDEX]);
+        }
+    }
+
+    function _send(message, callback, connect = false) {
+        local headers = {
+            "Content-Type" : "application/json;charset=UTF-8"
+        }
+
+        foreach (k, v in _configuration.requestHeaders) {
+            headers[k] <- v;
+        }
+
+        foreach (name, cookie in _cookies) {
+            if (!("Cookie" in headers)) {
+                headers["Cookie"] <- "";
+            }
+            headers["Cookie"] += name + "=" + cookie + "; ";
+        }
+
+        local jsonMsg = null;
+        try {
+            jsonMsg = http.jsonencode(message);
+        } catch (e) {
+            _logError("Cannot convert message to JSON string: " + e);
+            return;
+        }
+
+        _log("Sending message: " + jsonMsg);
+
+        local httpRequest = http.post(_configuration.url, headers, jsonMsg);
+        if (connect) {
+            _connectHttpRequest = httpRequest;
+        } else {
+            _extraHttpRequest = [httpRequest, callback];
+        }
+
+        local sent = function(response) {
+            _onSent(response, callback, connect);
+        }.bindenv(this);
+
+        local timeout = _configuration.requestTimeout;
+        if (connect) {
+            // _advice.timeout is in msec, so we need to convert it to sec
+            timeout += _advice.timeout / 1000.0;
+        }
+
+        httpRequest.sendasync(sent, null, timeout);
+    }
+
+    function _onSent(response, callback, connect) {
+        if (connect) {
+            _connectHttpRequest = null;
+        } else {
+            _extraHttpRequest = null;
+        }
+
+        _log("Response received:");
+        _logTable(response);
+
+        if (!_statusIsOk(response.statuscode)) {
+            callback(Bayeux.Error(BC_ERROR_TYPE.TRANSPORT_FAILED, response.statuscode));
+            _processQueue();
+            return;
+        }
+
+        _updateCookies(response.rawheaders);
+
+        local tableMsg = null;
+        try {
+            tableMsg = http.jsondecode(response.body);
+        } catch (e) {
+            _logError("Response body is not a valid JSON: " + e);
+            _logError("Message: " + response.body);
+            callback(Bayeux.Error(BC_ERROR_TYPE.BAYEUX_ERROR, "Response body is not a valid JSON: " + e));
+            _processQueue();
+            return;
+        }
+
+        if (tableMsg != null) {
+            callback(null);
+            _handleMessages(tableMsg);
+        }
+
+        _processQueue();
+    }
+
+    function _updateCookies(headers) {
```

Comment:
Just curious, should we ever clean cookies up?

Reply:
We clean them after every disconnection. But there might be some cases where we should also clean them. Added a TODO on it.